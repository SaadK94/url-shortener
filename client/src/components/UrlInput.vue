<template>
  <form @submit.prevent="shortenUrl">
    <input
      type="text"
      placeholder="Paste your URL here"
      :class="invalidUrl ? 'error' : ''"
      v-model="url"
      data-jest-input
    />
    <button
      type="submit"
      :class="!url ? 'disabled' : ''"
      :disabled="!url"
      data-jest-button
    >
      Shorten
    </button>
  </form>
  <p class="error" v-if="invalidUrl" data-jest-message>
    Invalid URL entered, please provide a valid URL
  </p>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import { mapActions } from "vuex";

export default defineComponent({
  data() {
    return {
      url: "",
      invalidUrl: false,
    };
  },
  watch: {
    url() {
      this.invalidUrl = false;
    },
  },
  methods: {
    ...mapActions(["submitUrl"]),
    isValidUrl(str: string) {
      try {
        new URL(str);
        return true;
      } catch (_) {
        return false;
      }
    },
    async shortenUrl() {
      const isValidUrl = this.isValidUrl(this.url);
      if (isValidUrl) {
        this.invalidUrl = false;
        await this.submitUrl(this.url);
        this.url = "";
      } else {
        this.invalidUrl = true;
      }
    },
  },
});
</script>

<style scoped>
form {
  display: flex;
}

input {
  padding: 12px 16px;
  border-radius: 8px;
  background-color: #f4f5f5;
  border: none;
  font-size: 20px;
  margin-right: 16px;
  flex-grow: 1;
}

input.error {
  background-color: rgba(241, 15, 1, 0.12);
  color: #f10f01;
}

input:focus {
  outline: none;
}

p.error {
  color: #f10f01;
}

button {
  padding: 16px;
  border-radius: 8px;
  cursor: pointer;
  color: white;
  font-size: 20px;
  text-align: center;
  background-color: #27515f;
  width: 175px;
}

button:hover {
  background-color: #048b97;
}

button.disabled {
  background-color: rgba(0, 0, 0, 0.12);
  color: rgba(0, 0, 0, 0.38);
  cursor: default;
}

@media (max-width: 576px) {
  form {
    flex-direction: column;
    align-items: stretch;
  }

  input {
    margin: 0 0 10px 0;
  }

  button {
    width: 100%;
    margin-bottom: 12px;
  }
}
</style>
