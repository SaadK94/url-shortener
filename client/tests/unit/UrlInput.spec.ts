import { shallowMount, VueWrapper } from "@vue/test-utils";
import UrlInput from "@/components/UrlInput.vue";

describe("UrlInput.vue", () => {
  let wrapper: VueWrapper<any>;

  beforeEach(() => {
    wrapper = shallowMount(UrlInput);
  });

  it("adds the error class when there is an invalid url", async () => {
    expect(wrapper.find("[data-jest-input]").classes()).not.toContain("error");
    await wrapper.setData({ invalidUrl: true });
    expect(wrapper.find("[data-jest-input]").classes()).toContain("error");
  });

  it("shows the error message when there is an invalid url", async () => {
    expect(wrapper.find("[data-jest-message]").exists()).toBe(false);
    await wrapper.setData({ invalidUrl: true });
    expect(wrapper.find("[data-jest-message]").exists()).toBe(true);
  });

  it("disables the button when the input is empty", async () => {
    expect(wrapper.find("[data-jest-button]").attributes()).toHaveProperty(
      "disabled"
    );
    await wrapper.setData({ url: "https://www.google.co.uk" });
    expect(wrapper.find("[data-jest-button]").attributes()).not.toHaveProperty(
      "disabled"
    );
  });

  it("should only submit the form with a url present", async () => {
    wrapper.vm.shortenUrl = jest.fn();
    await wrapper.find("[data-jest-button]").trigger("submit");
    expect(wrapper.vm.shortenUrl).not.toHaveBeenCalled();
    await wrapper.setData({ url: "https://www.google.co.uk" });
    await wrapper.find("[data-jest-button]").trigger("submit");
    expect(wrapper.vm.shortenUrl).toHaveBeenCalled();
  });

  it("matches the snapshot", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
