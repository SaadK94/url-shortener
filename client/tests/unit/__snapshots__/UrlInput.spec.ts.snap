// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UrlInput.vue matches the snapshot 1`] = `
<div
  data-v-app=""
>
  
  <form>
    <input
      class=""
      data-jest-input=""
      placeholder="Paste your URL here"
      type="text"
    />
    <button
      class="disabled"
      data-jest-button=""
      disabled=""
      type="submit"
    >
       Shorten 
    </button>
  </form>
  <!--v-if-->
  
</div>
`;
